[BE] refactor: 로깅 관련 리팩토링 및 콘솔 로깅 방식 변경 #339

yuyoungrhee opened the pull request:
<!-- PR 제목 예시
ex. [BE] 로그인 API 구현
ex. [AN] 페이지 구현 -->

## 📌 변경 내용 & 이유
<!-- ex. 
- 사용자 로그인 기능 구현
- 로그인 실패 시 에러 메시지 반환 처리 추가 -->

### 1. 콘솔 로그 가독성 개선

- Log4j2.xml 설정을 수정하여 <Console> Appender의 레이아웃을` <JsonLayout>`에서` <PatternLayout>`으로 변경했습니다.
    - 콘솔에는 텍스트 패턴 로그가 출력
    - 파일에는 기존과 동일하게 JSON 형식의 로그가 출력
- consoleLogger를 추가하여 console을 위한 로깅을 추가했습니다.

<img width="1278" height="285" alt="image" src="https://github.com/user-attachments/assets/0acbf20f-9946-4b8a-82b8-4f56cba9b27d" />

### 2. 전체 요청 로깅 기능 추가
- try-finally 구조 적용: chain.doFilter 호출 시 예외가 발생하더라도 finally 블록이 실행되어, 예외 발생 여부와 관계없이 모든 요청의 로그를 남길 수 있습니다.

- 필터 실행 순서 보장: RequestLoggingFilter에 @Order(Ordered.HIGHEST_PRECEDENCE + 1)을 명시적으로 부여하여, MdcLoggingFilter 다음으로 실행되도록 순서를 보장합니다.

### 3. 마스킹 로직 리팩토링
- 기존 JsonMaskingPrettyFormatter는 "password", "url"과 같은 특정 필드 이름을 하드코딩하여 마스킹을 적용

- MaskingSupport 클래스는 @Masking 커스텀 어노테이션이 붙어있는 필드를 찾아 마스킹합니다.
    - 선언적 마스킹: 마스킹이 필요한 DTO 필드에 @Masking 어노테이션을 붙이기만 하면 됨
    - 낮은 결합도: 마스킹 로직이 필드 이름이 아닌 어노테이션에 의존하게 되어 유지보수성이 크게 향상

- RecordComponent와 Field에서 모두 어노테이션을 탐색하는 로직을 구현했습니다.
    - Log에 사용되는 일부 dto는 일반class / reqeust&resposne dto는 record를 사용하므로 둘 다 탐색해야됨

## 🧪 테스트 방법 (선택)
<!-- 코드 리뷰하는 팀원이 돌려볼 수 있도록 작성해주기 
ex. ReservationTest의 XXXX 테스트메서드로 확인 
ex. 어떤화면의 어떤기능에서 확인 -->

``` xml
        <!-- JSON 파일 Logger-->
        <Logger name="jsonLogger" level="info" additivity="false">
            <AppenderRef ref="RollingFile"/>
            <AppenderRef ref="Console"/> <--이거 추가: jsonLogger도 console에서 찍겠다.
        </Logger>
```
log4j2.xml에 이거 한 줄 추가해주면 console로그 + json로그 둘 다 찍힘 json로그만보고싶으면 consoleLogger에서 AppenderRef 삭제하면됨
콘솔에서 열심히 돌려봐주세요...... 오류있으면 말해주삼

## 📢 논의하고 싶은 내용
<!-- 이러이러해서 이부분을 이러이러하게 구현했는데 이방식이 괜찮은지? -->


## 🧩 관련 이슈
<!-- 이슈 태그: #123 -->
<!-- 이슈 닫기: close #123 -->
#339 


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- 신기능
  - 비밀번호·토큰·URL 등 민감정보 마스킹 강화로 로그 개인정보 보호 향상
  - 콘솔용/JSON용 로그 분리 출력 및 요청/응답 로그 가독성 개선
  - 필수 파라미터 누락 시 명확한 한글 오류 메시지 제공
- 리팩터링
  - 로깅 체계 전면 개편(AOP/필터 기반 요청·응답·예외 로깅 정비)
  - 관리자 엔드포인트는 AOP 로깅에서 제외하여 불필요한 노이즈 감소
- 잡무
  - 로그 필터 등록 및 설정 정리, 구성 클래스 명칭/패키지 정돈
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

yuyoungrhee committed:
[BE] chore: MdcLoggingFilter를 서블릿필터체인에 등록하도록 변경 #339

- MdcLoggingFilter를 @Component로 스프링 필터 체인에 등록하는 대신 FilterRegistrationBean으로 서블릿 필터 체인에 등록: 우선순위를 최상위로 설정해 모든 요청에 대해 MDC를 정상적으로 세팅하도록 변경한 작업

yuyoungrhee committed:
[BE] feat: 모니터링을 위한 예외 httpStatus code 추가 로깅 #339

yuyoungrhee committed:
[BE] feat: S3 url 로깅 시 마스킹 추가 #339

yuyoungrhee committed:
[BE] refactor: 로깅 시 마스킹 방법을 커스텀 어노테이션 이용으로 변경 #339

- 기존 마스킹 로직 하드코딩에서 어노테이션 선언 방식으로 개선

yuyoungrhee committed:
[BE] feat: Filter까지만 오는 Request도 로깅 추가 #339

- 기존: AOP를 통해 Controller까지 오는 Request들만 로깅
- 변경: RequestLoggingFilter를 통해 Filter까지만 오는 Request도 로깅할 수 있도록 추가

yuyoungrhee committed:
[BE] feat: console 로그 형식 추가 #339

yuyoungrhee commented:
@coderabbitai 로깅 관련 리뷰 달아줘

gabean13 commented:
@coderabbitai 레빗아 리뷰달아줘

yuyoungrhee committed:
[BE] feat: 필수 파라미터 누락 예외 400으로 처리 #339

- 필수 파라미터 누락 예외가 500으로 잡히는 문제 해결

yuyoungrhee committed:
[BE] feat: 관리자 로깅 제외 #339

- admin패키지는 AOP Logging 대상에서 제외

yuyoungrhee committed:
[BE] fix: consoleLogger에서 throwable이 없는 경우 NPE 방지 #339

yuyoungrhee committed:
[BE] fix: 필터 순서에서 같은 순위를 주고있는 문제 해결 #339

yuyoungrhee committed:
[BE] refactor: log 클래스 분리 및 패키지 구조 변경 #339

yuyoungrhee committed:
[BE] fix: MDC import 라이브러리 slf4j로 변경 #339

- 기존에 사용하던 JBoss Logging 프레임워크는 일반적인 Spring Boot 애플리케이션에서는 거의 사용하지 않음

yuyoungrhee committed:
[BE] fix: 마스킹 과정에서 문자열 길이 검증 추가로 인덱스범위예외 가능성 제거 #339

yuyoungrhee committed:
[BE] refactor: ResponseLog 필드 의미 불일치 변경 #339

yuyoungrhee commented on backend/src/main/java/com/onair/hearit/common/log/dto/ExceptionLog.java:
```diff
@@ -14,23 +14,34 @@ public class ExceptionLog {
+
+        public static Status from(HttpStatus status) {
+            return new Status(status.value(), status.name());
+        }
```
HttpStatus가 지금 null일수가있나요?

yuyoungrhee commented on backend/src/main/java/com/onair/hearit/common/log/dto/ResponseLog.java:
```diff
@@ -1,25 +1,25 @@
             long timeTakenMs) {
-        return new ResponseLog(
+        return new ResponseLog<>(
                 "RESPONSE",
                 timestamp.toString(),
                 requestInfo,
-                responseEntity,
+                responseEntity.getBody(),
                 timeTakenMs);
```
ResponseEntity가 지금 null일수가있나요?

yuyoungrhee commented on backend/src/main/java/com/onair/hearit/common/log/formatter/ConsoleLogFormatter.java:
```diff
@@ -0,0 +1,58 @@
+        );
+    }
+
+    private static String toFlatParamString(Map<String, List<String>> params) {
+        return params.entrySet().stream()
+                .map(entry -> entry.getKey() + "=" + entry.getValue())
+                .collect(Collectors.joining(", ", "{", "}"));
+    }
```
콘솔은 그냥해도되지않나요?

yuyoungrhee commented on backend/src/main/resources/log4j2/log4j2.xml:
```diff
@@ -87,6 +81,16 @@
             <AppenderRef ref="RollingFile"/>
         </Root>
 
+        <!-- 콘솔 Logger-->
+        <Logger name="consoleLogger" level="info" additivity="false">
+            <AppenderRef ref="Console"/>
+        </Logger>
+
+        <!-- JSON 파일 Logger-->
+        <Logger name="jsonLogger" level="info" additivity="false">
+            <AppenderRef ref="RollingFile"/>
+        </Logger>
+
```
`@slf4j` 등 클래스 로그는 콘솔·파일에 중복으로 남아도 문제없지않나요?

yuyoungrhee commented on backend/src/main/java/com/onair/hearit/auth/dto/request/TokenReissueRequest.java:
```diff
@@ -1,6 +1,10 @@
+import com.onair.hearit.common.log.mask.Masking;
+import com.onair.hearit.common.log.mask.MaskingType;
+
 public record TokenReissueRequest(
+        @Masking(type = MaskingType.TOKEN)
         String refreshToken
 ) {
```
FIlterRequest에서는 컨트롤러에 못 도달한 요청 추적 목적으로, 헤더와 바디는 로깅 대상에 포함되지 않습니다. 파라미터로 들어오는 것들은 민감정보가 아니므로 RequestLoggingFilter에서는 굳이 마스킹이 필요 없는 거 같은데, 맞나요

yuyoungrhee commented on backend/src/main/java/com/onair/hearit/common/log/mask/MaskingSupport.java:
```diff
@@ -0,0 +1,132 @@
+    private final ObjectMapper objectMapper;
+    private final Map<MaskingType, MaskingStrategy> maskingStrategyMap;
+
+    public Object mask(Object obj) {
+        try {
+            if (obj instanceof ResponseEntity<?> responseEntity) {
+                Object maskedBody = mask(responseEntity.getBody());
+                return ResponseEntity.status(responseEntity.getStatusCode())
+                        .headers(responseEntity.getHeaders())
+                        .body(maskedBody);
+            }
+
+            JsonNode rootNode = objectMapper.valueToTree(obj);
+            applyMaskingWithAnnotations(obj, rootNode);
+            return rootNode;
+        } catch (Exception e) {
+            return "[마스킹 실패: " + e.getMessage() + "]";
+        }
+    }
```
이거 일단 냅둘게요 좀바꿨더니 masking을 잘 못 찾아서 성능 리팩터링 뭐 나중에 비동기로바꾸든지 해서 같이 바꾸면 좋긴할듯

JO-eusan commented on backend/src/main/java/com/onair/hearit/auth/dto/request/KakaoLoginRequest.java:
```diff
@@ -1,6 +1,10 @@
+import com.onair.hearit.common.log.mask.MaskingType;
+
 public record KakaoLoginRequest(
+        @Masking(type = MaskingType.TOKEN)
```
`@Masking(type = MaskingType.TOKEN) String accessToken` record에서는 이렇게 개행 안넣는거 어떤가요?

JO-eusan commented on backend/src/main/java/com/onair/hearit/auth/dto/request/SignupRequest.java:
```diff
@@ -1,8 +1,15 @@
 public record SignupRequest(
+
         String localId,
+
```
이 부분도 꼭 개행을 넣지 않아도 될 것 같아요.

yuyoungrhee committed:
[BE] feat: console 로깅 시에도 민감정보 마스킹 추가 #339

JO-eusan commented on backend/src/main/java/com/onair/hearit/auth/dto/response/LoginTokenResponse.java:
```diff
@@ -1,7 +1,14 @@
+
 public record LoginTokenResponse(
+
+        @Masking(type = MaskingType.TOKEN)
```
ResponseDto도 마찬가지로 모두 개행을 넣지 않아도 될 것 같아요.

yuyoungrhee committed:
[BE] fix: log 관련 발생할 수 있는 NPE 방지용 null체크 추가 #339

JO-eusan commented on backend/src/main/java/com/onair/hearit/common/log/config/LogFilterConfig.java:
```diff
@@ -0,0 +1,28 @@
+import org.springframework.core.Ordered;
+
+@Configuration
+public class LogFilterConfig {
```
우선순위를 Config에서 한 번에 볼 수 있어서 놓은데여? 😃

yuyoungrhee committed:
[BE] refactor: 클래스명을 로그 메시지 설정 역할에 맞게 LogMessageConfig로 변경 #339

JO-eusan commented on backend/src/main/java/com/onair/hearit/common/log/exception/FilterExceptionLogger.java:
```diff
@@ -1,32 +1,31 @@
 
-    private final JsonMaskingPrettyFormatter jsonMaskingPrettyFormatter;
     private final Logger errorLogger = LogManager.getLogger("errorLogger");
+    private static final Logger consoleLogger = LogManager.getLogger("consoleLogger");
```
static 필드라면 최상단에 두고 개행을 다른 필드가 나오기 전에 추가하는거 어떨까요?

gabean13 replied:
errorLogger도 static 껴주3

JO-eusan commented on backend/src/main/java/com/onair/hearit/common/log/exception/FilterExceptionLogger.java:
```diff
@@ -63,6 +68,14 @@ public void warn(ProblemDetail problemDetail) {
 
-        log.warn(jsonMaskingPrettyFormatter.convertToPrettyJson(exceptionLog));
+        jsonLogger.warn(maskingSupport.mask(exceptionLog));
+        consoleLogger.warn("[CLIENT ERROR] {} {} from {} → status: {} / title: {} / detail: {}",
```
warn이면 client 에러로 찍고 있는건가요? 
client error가 아닌 경우에 warn일 경우는 없을까요..

yuyoungrhee replied:
동일하게 WARN으로 찍도록 변경했습니다.

JO-eusan commented on backend/src/main/java/com/onair/hearit/common/log/formatter/ConsoleLogFormatter.java:
```diff
@@ -0,0 +1,73 @@
+
+    private static String toFlatParamString(Map<String, List<String>> params) {
+        return params.entrySet().stream()
+                .map(entry -> entry.getKey() + "=" + entry.getValue())
```
"생ㅜ"가 뭐에여?

JO-eusan commented on backend/src/main/java/com/onair/hearit/common/log/mask/strategy/EdgeMaskingStrategy.java:
```diff
@@ -0,0 +1,21 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class EdgeMaskingStrategy implements MaskingStrategy {
```
개행 추가해주세요

JO-eusan commented on backend/src/main/java/com/onair/hearit/common/log/mask/strategy/TokenMaskingStrategy.java:
```diff
@@ -0,0 +1,23 @@
+
+@Component
+public class TokenMaskingStrategy implements MaskingStrategy {
+    private static final int PREFIX_VISIBLE_LENGTH = 5; //토큰 유형, 발급처 또는 서명 방식 파악용
```
개행~

JO-eusan commented on backend/src/main/java/com/onair/hearit/common/log/mask/strategy/MaskingStrategy.java:
```diff
@@ -0,0 +1,9 @@
+import com.onair.hearit.common.log.mask.MaskingType;
+
+public interface MaskingStrategy {
+    MaskingType type();
```
개행~~

JO-eusan commented on backend/src/main/java/com/onair/hearit/common/log/mask/strategy/UrlMaskingStrategy.java:
```diff
@@ -0,0 +1,23 @@
+
+@Component
+@RequiredArgsConstructor
+public class UrlMaskingStrategy implements MaskingStrategy {
```
FileUrlMaskingStrategy는 어떨까요?

JO-eusan commented on backend/src/main/java/com/onair/hearit/common/log/LoggingAspect.java:
```diff
@@ -155,21 +175,43 @@ private Optional<Throwable> extractThrowableFromArgs(Object[] args) {
                 errorDetail);
-        log.warn(jsonMaskingPrettyFormatter.convertToPrettyJson(exceptionLog));
+        jsonLogger.warn(maskingSupport.mask(exceptionLog));
+        consoleLogger.warn(
```
이것도 warn이 client error 인가여?

JO-eusan commented on backend/src/main/java/com/onair/hearit/dto/response/OriginalAudioResponse.java:
```diff
@@ -1,7 +1,13 @@
 public record OriginalAudioResponse(
+
         Long id,
+
```
개행 지워도 되지 않을까여, 다른 Response도요!

JO-eusan reviewed (changes requested):
멍구.. 로깅 로직 복잡한데 짱이에여.
몇 가지 커멘트 남겨두었습니다!

Byesol commented on backend/src/main/java/com/onair/hearit/common/exception/GlobalExceptionHandler.java:
```diff
@@ -43,6 +44,11 @@ public ProblemDetail handleTypeMismatch(MethodArgumentTypeMismatchException ex,
         return buildProblemDetail(ErrorCode.INVALID_INPUT, "잘못된 파라미터 값: " + ex.getValue(), request);
     }
 
+    @ExceptionHandler(MissingServletRequestParameterException.class)
```
오 오류 처리 좋아요

Byesol reviewed (commented):
<img width="2698" height="52" alt="image" src="https://github.com/user-attachments/assets/0957a1f4-06e0-4201-9cd2-8b4f71918012" />
현재 관리자 페이지에서 에러를 발생시키면 로깅에 admin 관련 로깅이 찍히고 있어요

gabean13 commented on backend/src/main/java/com/onair/hearit/common/log/LoggingAspect.java:
```diff
@@ -35,8 +38,10 @@
-    private final JsonMaskingPrettyFormatter jsonMaskingPrettyFormatter;
     private final Logger errorLogger = LogManager.getLogger("errorLogger");
+    private static final Logger consoleLogger = LogManager.getLogger("consoleLogger");
+    private static final Logger jsonLogger = LogManager.getLogger("jsonLogger");
```
errorLogger도 static 끼워주3

gabean13 reviewed (commented):
멍구 고생했어용

yuyoungrhee committed:
[BE] refactor: 개행, 일관성 등 컨벤션 반영 #339

yuyoungrhee committed:
[BE] refactor: 콘솔 로깅 포맷 변경 #339

- REQUEST 로그 양옆에 따옴표 붙는 오류 해결
- logType이 콘솔에서도 그대로 찍히도록 변경

yuyoungrhee committed:
[BE] fix: 불필요한 로깅이 찍히지 않도록 수정 #339

- RequestLoggingFilter에서 브라우저에서 발생하는 불필요한 request(/favicon.ico 등)은 로깅이 찍히지 않도록 수정

yuyoungrhee commented on backend/src/main/java/com/onair/hearit/common/log/formatter/ConsoleLogFormatter.java:
```diff
@@ -0,0 +1,73 @@
+        return masked.length() > 200 ? masked.substring(0, 200) + "...(생략)" : masked;
+    }
+
+    private static String maskBodySensitiveData(String raw) {
+        for (String key : SENSITIVE_KEYS) {
+            // record toString style: ClassName[field1=value1, key=value2, ...]
+            String regex = String.format("(?i)(%s=)([^,\\]]+)", key);
+            raw = raw.replaceAll(regex, "$1****");
+        }
+        return raw;
+    }
```
지금도잘돼서 나중에문제생기면 바꿀게욘

yuyoungrhee committed:
[BE] feat: 예외 이름도 함께 로깅하도록 추가 #339

JO-eusan reviewed (approved):
멍구 리뷰 반영 확인했습니다. Approve 할게요!

gabean13 reviewed (approved):
고생햇3
